File: two_sum_iv_input_is_a_bst.c
// https://leetcode.com/problems/two-sum-iv-input-is-a-bst/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *find(struct TreeNode *root, int k) {
  if (!root) {
    return NULL;
  }

  if (root->val == k) {
    return root;
  }

  if (root->val < k) {
    return find(root->right, k);
  }

  return find(root->left, k);
}

int helper(struct TreeNode *t, struct TreeNode *root, int k) {
  if (!t) {
    return 0;
  }

  int target = find(root, k - t->val);
  if (target && target != t) {
    return 1;
  }

  return helper(t->left, root, k) || helper(t->right, root, k);
}
bool findTarget(struct TreeNode *root, int k) { return helper(root, root, k); }

File: algorithms/easy/two_sum_iv_input_is_a_bst.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec find_target(root :: TreeNode.t() | nil, k :: integer) :: boolean
  def find_target(root, k) do
    dfs(root, MapSet.new(), k)
    |> elem(0)
  end

  def dfs(nil, set, _k), do: {false, set}

  def dfs(curr, set, k) do
    if (k - curr.val) in set do
      {true, set}
    else
      {r, set} = dfs(curr.left, MapSet.put(set, curr.val), k)

      if r do
        {r, set}
      else
        dfs(curr.right, set, k)
      end
    end
  end
end
